### ISSUE TITLE
compile cosmic-applets individually?

### ISSUE THREAD

**Issue by MrMonotone** [lilyinstarlight/nixos-cosmic#34]
Would it be possible to have each applet build as it own job to lower the ram requirements when compiling locally? This would make it convenient to add flags for the ones that you want as well.

**Comment by lilyinstarlight** [comment]
Yeah this could probably technically be done! Just curious, are you building the applets yourself because you aren't using the binary cache? (e.g. all binaries here should be cached for x86_64-linux and aarch64-linux on cosmic.cachix.org and once this is merged back into nixpkgs, cache.nixos.org)

**Comment by MrMonotone** [comment]
The cache works great. I just tried to build without it and I ran out of ram with 32GB of it. I tried limiting the jobs to 1 and still ran out. It is not a priority but it would be nice for someone who wants to build locally.

**Comment by lilyinstarlight** [comment]
> The cache works great. I just tried to build without it and I ran out of ram with 32GB of it. I tried limiting the jobs to 1 and still ran out. It is not a priority but it would be nice for someone who wants to build locally.

Limiting jobs likely won't help, but using `--cores 1` or some similar number to reduce build parallelism probably will help

But either way, making the derivation take a list argument or something to define which applets to build is probably doable

**Comment by MrMonotone** [comment]
> But either way, making the derivation take a list argument or something to define which applets to build is probably doable
Would it be possible to build each as an individual job?

**Comment by lilyinstarlight** [comment]
Possibly? Idk, it would be a lot more wasted compiling (especially compared to `--cores 1`). But I'm not too opposed

**Comment by atagen** [comment]
had same issue testing #64 . I'm for whatever helps the cache work best since 99% of the time that's what I'll be using too, but it was an unpleasant surprise to OOM while testing a one line nix-side change :smile:

**Comment by lilyinstarlight** [comment]
> had same issue testing #64 . I'm for whatever helps the cache work best since 99% of the time that's what I'll be using too, but it was an unpleasant surprise to OOM while testing a one line nix-side change 😄

I'm still not too opposed to doing thiz. It's just a tradeoff between wasting shared compilation unit by splitting it up versus just setting a parallelism amount (`--cores [n]`) that won't OOM for however much RAM you do have on your system (which is just a general Nix thing not necessarily specific to Rust or COSMIC, e.g. highly templated C++ apps do it too which is why they actually have a separate set of systems in Hydra with tons of RAM to build some derivations with "big-parallel" set)

Alternatively we could just disable parallel compilation just for cosmic-applets too in the derivation, which would solve the OOM problem and just be a little slower for people who do have enough RAM (but tbh not much slower, so it's definitely viable)

**Comment by m-kuzmin** [comment]
> But either way, making the derivation take a list argument or something to define which applets to build is probably doable

Could we use separate outputs like `<drv>.share` or `.doc` (but replaced with applet names) to separate what gets built?

**Comment by lilyinstarlight** [comment]
> Could we use separate outputs like `<drv>.share` or `.doc` (but replaced with applet names) to separate what gets built?

If you have some motivation that is worth the somewhat increased maintenance cost maybe. What would that help you do?

**Comment by m-kuzmin** [comment]
It was just an implementation idea, I don't really need to select applets.

**Comment by lilyinstarlight** [comment]
> It was just an implementation idea, I don't really need to select applets.

To reduce confusion, I'll just close this for now unless someone brings forward a need for this :)

(Also separate outputs does not change compilation and would not help the original request)
